Improve message when no results are found and hide load more button

lcezermf opened the pull request:
## Why is this change being made?

It fixes #35 by adding a friendly message about no results when the filter does not return data

## How is this being accomplished?

- Add a message when no results are returned
- Change from streams to assign
- Hide load more button when no need to show

### Testing

- Run project `iex -S mix phx.server`
- Access http://localhost:4000/en/courses
- Apply some filter
- [ ] Must return results
- Add a search like "wwwwww" and try to filter
- [ ] Must return "no results" message


https://github.com/danielbergholz/techschool.dev/assets/2160392/5e700e29-dd4a-4e1b-8d2d-cb60a1c05280

lcezermf committed:
Improve message when no results are found and hide load more button

- Add a message when no results are returned-
- Change from streams to assign
- Hide load more button when no need to show

danielbergholz commented on lib/techschool_web/live/course_live/index.html.heex:
```diff
@@ -9,11 +9,16 @@
-    <div class="flex flex-wrap justify-center gap-8 md:gap-10" id="courses" phx-update="stream">
-      <.course_card :for={{dom_id, course} <- @streams.courses} id={dom_id} course={course} />
+    <div class="flex flex-wrap justify-center gap-8 md:gap-10" id="courses">
+      <p :if={Enum.empty?(@courses)} id="no-results">
```
A small suggestion to avoid adding business logic inside a component: instead of calling `Enum.empty` here inside the component, you can add a new variable inside the `handle_event` and `handle_params` called `has_courses` or `no_results` for example, and assign it to the WebSocket. 

This way the code would look like this:

```suggestion
      <p :if={@no_results} id="no-results">
```

Also, by doing this, having the courses inside the `@streams` shouldn't be a problem anymore, since you are counting the courses inside the `handle_event` and `handle_params`, and not inside the component

lcezermf replied:
Sure I will check this and the stream back, thanks for the feedback!

danielbergholz commented on test/techschool_web/live/course_live_test.exs:
```diff
@@ -13,4 +13,36 @@ defmodule TechschoolWeb.CourseLiveTest do
     end
   end
+
+  describe "GET /:locale/courses with search query string" do
```
Thank you for writing tests! 🙌

danielbergholz reviewed (commented):
Thank you for the video and very descriptive PR! 🚀

After taking a look on this [blog post](https://fly.io/phoenix-files/phoenix-dev-blog-streams/#should-streams-be-used-by-default-now-for-lists-of-items) and some other resources, I think it's better to keep using `stream` instead of `assign`

lcezermf committed:
Create variable has_courses to be used on html

The variable name follow the has_* pattern as has_more_courses_to_load and
check if there are courses returned by the query, that is why the name is
affirmative and I use !Enum.empty?/1 to check the value.

lcezermf committed:
Add streams back to hold the query results and build UI properly

danielbergholz commented on lib/techschool_web/live/course_live/index.ex:
```diff
@@ -69,7 +70,8 @@ defmodule TechschoolWeb.CourseLive.Index do
     |> assign(:has_more_courses_to_load, has_more_courses_to_load)
-    |> stream(:courses, courses)
+    |> assign(:has_courses, !Enum.empty?(courses))
+    |> stream(:courses, courses, reset: true)
```
Here, since we are loading more courses into the screen, I don't want to reset the other ones already loaded

```suggestion
    |> assign(:has_courses, !Enum.empty?(courses))
    |> stream(:courses, courses)
```

lcezermf replied:
I tried here, if I remove the :reset, once I get to the first select and filter something, the URL gets updated but not the records on the page. Are you sure it needs to be removed?!

danielbergholz replied:
Really? I tested locally now and everything seems to be working normally 🤔
You just need to remove the `reset: true` inside the `handle_event("load_more")` and keep the one inside `handle_params`

lcezermf replied:
Pushed! d12bede5d04c01b4dd252da41ec64f7ece113bf5

danielbergholz commented on lib/techschool_web/live/course_live/index.html.heex:
```diff
@@ -9,11 +9,20 @@
+      <.button phx-click="load_more" class="mt-8" disabled={@has_more_courses_to_load == false}>
+        <%= gettext("Load More") %>
+      </.button>
+    <% end %>
```
I think we can be less verbose, and use the `:if` inside the component. Also, feel free to remove the `disabled`

```suggestion
    <.button phx-click="load_more" class="mt-8" :if={@has_courses and @has_more_courses_to_load}>
      <%= gettext("Load More") %>
    </.button>
```

lcezermf replied:
Nice! I work on a elixir/phoenix project that is not updated to the new version, so these :if/:for attributes are totally new to me, thanks for the advice!

danielbergholz reviewed (commented):
Sorry for the late reply!

lcezermf committed:
Improve button to load more using :if

lcezermf committed:
Remove reset: true

danielbergholz reviewed (approved):
This is great work @lcezermf! Thank you for the contribution 🚀🔥
